### PR TITLE
expand 0 length edge using var len traversal

### LINKED ISSUE #2836: Crash on Query
Hi there, redisgraph built from commit c425ad8d9a9009aa362dac358a2431cdb1dc2d05 will crash when running the following query:

```
GRAPH.QUERY a "MATCH (), path_0 = ()--()-->()--(node_3:label4), path_1 = ()-[edge_1*]-()<-[reltype7*0]-(node_3:labelunk) WHERE TRUE >= edge_1.a UNWIND [] AS z RETURN edge_1.d"
```

Crash log from the server:

```
2590711:C 21 Jan 2023 03:45:47.326 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
2590711:C 21 Jan 2023 03:45:47.326 # Redis version=255.255.255, bits=64, commit=c95ff0f3, modified=0, pid=2590711, just started
2590711:C 21 Jan 2023 03:45:47.326 # Configuration loaded
2590711:M 21 Jan 2023 03:45:47.328 * monotonic clock: POSIX clock_gettime
                _._
           _.-``__ ''-._
      _.-``    `.  `_.  ''-._           Redis 255.255.255 (c95ff0f3/0) 64 bit
  .-`` .-```.  ```\/    _.,_ ''-._
 (    '      ,       .-`  | `,    )     Running in standalone mode
 |`-._`-...-` __...-.``-._|'` _.-'|     Port: 6379
 |    `-._   `._    /     _.-'    |     PID: 2590711
  `-._    `-._  `-./  _.-'    _.-'
 |`-._`-._    `-.__.-'    _.-'_.-'|
 |    `-._`-._        _.-'_.-'    |           https://redis.io
  `-._    `-._`-.__.-'_.-'    _.-'
 |`-._`-._    `-.__.-'    _.-'_.-'|
 |    `-._`-._        _.-'_.-'    |
  `-._    `-._`-.__.-'_.-'    _.-'
      `-._    `-.__.-'    _.-'
          `-._        _.-'
              `-.__.-'

2590711:M 21 Jan 2023 03:45:47.334 * <graph> Starting up RedisGraph version 99.99.99.
2590711:M 21 Jan 2023 03:45:47.339 * <graph> Thread pool created, using 72 threads.
2590711:M 21 Jan 2023 03:45:47.339 * <graph> Maximum number of OpenMP threads set to 72
2590711:M 21 Jan 2023 03:45:47.339 * Module 'graph' loaded from /RedisGraphSrc/bin/linux-x64-release/src/redisgraph.so
2590711:M 21 Jan 2023 03:45:47.341 # Server initialized
2590711:M 21 Jan 2023 03:45:47.341 # WARNING Memory overcommit must be enabled! Without it, a background save or replication may fail under low memory condition. Being disabled, it can also cause failures without low memory condition, see https://github.com/jemalloc/jemalloc/issues/1328. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
2590711:M 21 Jan 2023 03:45:47.341 * Ready to accept connections tcp


=== REDIS BUG REPORT START: Cut & paste starting from here ===
2590711:M 21 Jan 2023 03:45:56.620 # Redis 255.255.255 crashed by signal: 11, si_code: 1
2590711:M 21 Jan 2023 03:45:56.620 # Accessing address: 0x7ff8
2590711:M 21 Jan 2023 03:45:56.620 # Crashed running the instruction at: 0x7ff87b833d5d

------ STACK TRACE ------
EIP:
/lib/x86_64-linux-gnu/libc.so.6(+0x1b1d5d)[0x7ff87b833d5d]

Backtrace:
/lib/x86_64-linux-gnu/libc.so.6(+0x42520)[0x7ff87b6c4520]
/lib/x86_64-linux-gnu/libc.so.6(+0x1b1d5d)[0x7ff87b833d5d]
/RedisGraphSrc/bin/linux-x64-release/src/redisgraph.so(orderExpressions+0x2a2)[0x7ff8790991e2]
/RedisGraphSrc/bin/linux-x64-release/src/redisgraph.so(buildMatchOpTree+0x429)[0x7ff879083649]
/RedisGraphSrc/bin/linux-x64-release/src/redisgraph.so(ExecutionPlan_PopulateExecutionPlan+0x77)[0x7ff8790822f7]
/RedisGraphSrc/bin/linux-x64-release/src/redisgraph.so(NewExecutionPlan+0x129)[0x7ff879082469]
/RedisGraphSrc/bin/linux-x64-release/src/redisgraph.so(ExecutionCtx_FromQuery+0x127)[0x7ff87907cb57]
/RedisGraphSrc/bin/linux-x64-release/src/redisgraph.so(_query+0x68)[0x7ff87907c618]
/RedisGraphSrc/bin/linux-x64-release/src/redisgraph.so(+0x25630f)[0x7ff8790c730f]
/lib/x86_64-linux-gnu/libc.so.6(+0x94b43)[0x7ff87b716b43]
/lib/x86_64-linux-gnu/libc.so.6(+0x126a00)[0x7ff87b7a8a00]

------ REGISTERS ------
2590711:M 21 Jan 2023 03:45:56.625 #
RAX:0000000000000ff8 RBX:0000000000000003
RCX:0000000001010040 RDX:0000000000000000
RDI:00007ff848001998 RSI:0000000000007ff8
RBP:00007ff877e6c1b0 RSP:00007ff877e6c068
R8 :000000000000007c R9 :0000000000000000
R10:00007ff850009408 R11:00007ff87b400900
R12:0000000000007ff8 R13:0000000000000003
R14:00007ff85000ca10 R15:00007ff8500180fc
RIP:00007ff87b833d5d EFL:0000000000010246
CSGSFS:002b000000000033
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c077) -> 00007ff800000000
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c076) -> 00007ff85000c770
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c075) -> 0000000000000000
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c074) -> 00007ff85000cb28
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c073) -> 00007ff800000002
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c072) -> 00007ff85000c968
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c071) -> 00007ff800000002
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c070) -> 00007ff85000ca10
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c06f) -> 00007ff800000004
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c06e) -> 00007ff85000115c
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c06d) -> 00007ff850009408
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c06c) -> 00007ff85000ca10
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c06b) -> 00007ff85000c968
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c06a) -> 00007ff85000c770
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c069) -> 00007ff85000c818
2590711:M 21 Jan 2023 03:45:56.625 # (00007ff877e6c068) -> 00007ff8790991e2

------ INFO OUTPUT ------
# Server
redis_version:255.255.255
redis_git_sha1:c95ff0f3
redis_git_dirty:0
redis_build_id:e1489f779f6ba070
redis_mode:standalone
os:Linux 5.4.0-136-generic x86_64
arch_bits:64
monotonic_clock:POSIX clock_gettime
multiplexing_api:epoll
atomicvar_api:c11-builtin
gcc_version:11.3.0
process_id:2590711
process_supervised:no
run_id:7791bb51b7e384d3b36f947dac1e2feb919cbe7e
tcp_port:6379
server_time_usec:1674272756618040
uptime_in_seconds:9
uptime_in_days:0
hz:10
configured_hz:10
lru_clock:13328372
executable:/redis_src/./src/redis-server
config_file:/redis_src/./redis.conf
io_threads_active:0
listener0:name=tcp,bind=127.0.0.1,port=6379

# Clients
connected_clients:1
cluster_connections:0
maxclients:10000
client_recent_max_input_buffer:20480
client_recent_max_output_buffer:0
blocked_clients:1
tracking_clients:0
clients_in_timeout_table:0
total_blocking_keys:0
total_blocking_keys_on_nokey:0

# Memory
used_memory:1354432
used_memory_human:1.29M
used_memory_rss:13688832
used_memory_rss_human:13.05M
used_memory_peak:1354432
used_memory_peak_human:1.29M
used_memory_peak_perc:101.13%
used_memory_overhead:951688
used_memory_startup:949632
used_memory_dataset:402744
used_memory_dataset_perc:99.49%
allocator_allocated:1627344
allocator_active:1982464
allocator_resident:14327808
total_system_memory:134823161856
total_system_memory_human:125.56G
used_memory_lua:31744
used_memory_vm_eval:31744
used_memory_lua_human:31.00K
used_memory_scripts_eval:0
number_of_cached_scripts:0
number_of_functions:0
number_of_libraries:0
used_memory_vm_functions:32768
used_memory_vm_total:64512
used_memory_vm_total_human:63.00K
used_memory_functions:184
used_memory_scripts:184
used_memory_scripts_human:184B
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
allocator_frag_ratio:1.22
allocator_frag_bytes:355120
allocator_rss_ratio:7.23
allocator_rss_bytes:12345344
rss_overhead_ratio:0.96
rss_overhead_bytes:-638976
mem_fragmentation_ratio:14.02
mem_fragmentation_bytes:12712136
mem_not_counted_for_evict:0
mem_replication_backlog:0
mem_total_replication_buffers:0
mem_clients_slaves:0
mem_clients_normal:1800
mem_cluster_links:0
mem_aof_buffer:0
mem_allocator:jemalloc-5.2.1
active_defrag_running:0
lazyfree_pending_objects:0
lazyfreed_objects:0

# Persistence
loading:0
async_loading:0
current_cow_peak:0
current_cow_size:0
current_cow_size_age:0
current_fork_perc:0.00
current_save_keys_processed:0
current_save_keys_total:0
rdb_changes_since_last_save:0
rdb_bgsave_in_progress:0
rdb_last_save_time:1674272747
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
rdb_saves:0
rdb_last_cow_size:0
rdb_last_load_keys_expired:0
rdb_last_load_keys_loaded:0
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_rewrites:0
aof_rewrites_consecutive_failures:0
aof_last_write_status:ok
aof_last_cow_size:0
module_fork_in_progress:0
module_fork_last_cow_size:0

# Stats
total_connections_received:1
total_commands_processed:1
instantaneous_ops_per_sec:0
total_net_input_bytes:223
total_net_output_bytes:201072
total_net_repl_input_bytes:0
total_net_repl_output_bytes:0
instantaneous_input_kbps:0.00
instantaneous_output_kbps:0.00
instantaneous_input_repl_kbps:0.00
instantaneous_output_repl_kbps:0.00
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
expired_stale_perc:0.00
expired_time_cap_reached_count:0
expire_cycle_cpu_milliseconds:0
evicted_keys:0
evicted_clients:0
total_eviction_exceeded_time:0
current_eviction_exceeded_time:0
keyspace_hits:0
keyspace_misses:1
pubsub_channels:0
pubsub_patterns:0
pubsubshard_channels:0
latest_fork_usec:0
total_forks:0
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0
total_active_defrag_time:0
current_active_defrag_time:0
tracking_total_keys:0
tracking_total_items:0
tracking_total_prefixes:0
unexpected_error_replies:0
total_error_replies:0
dump_payload_sanitizations:0
total_reads_processed:2
total_writes_processed:3
io_threaded_reads_processed:0
io_threaded_writes_processed:0
reply_buffer_shrinks:1
reply_buffer_expands:0
acl_access_denied_auth:0
acl_access_denied_cmd:0
acl_access_denied_key:0
acl_access_denied_channel:0

# Replication
role:master
connected_slaves:0
master_failover_state:no-failover
master_replid:deb18d2defc9db1575b335ee4ba3673c57c6d3cf
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:0
second_repl_offset:-1
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:0.020624
used_cpu_user:0.035693
used_cpu_sys_children:0.000000
used_cpu_user_children:0.000000
used_cpu_sys_main_thread:0.000000
used_cpu_user_main_thread:0.007632

# Modules
module:name=graph,ver=999999,api=1,filters=0,usedby=[],using=[],options=[]

# Commandstats
cmdstat_command|docs:calls=1,usec=2599,usec_per_call=2599.00,rejected_calls=0,failed_calls=0

# Errorstats

# Latencystats
latency_percentiles_usec_command|docs:p50=2605.055,p99=2605.055,p99.9=2605.055

# Cluster
cluster_enabled:0

# Keyspace
db0:keys=1,expires=0,avg_ttl=0

------ CLIENT LIST OUTPUT ------
id=7 addr=127.0.0.1:46018 laddr=127.0.0.1:6379 fd=7 name= age=7 idle=0 flags=b db=0 sub=0 psub=0 ssub=0 multi=-1 qbuf=0 qbuf-free=20474 argv-mem=171 multi-mem=0 rbs=1024 rbp=0 obl=0 oll=0 omem=0 tot-mem=22467 events=r cmd=graph.QUERY user=default redir=-1 resp=2

------ MODULES INFO OUTPUT ------
# graph_executing commands
graph_command:GRAPH.QUERY MATCH (), path_0 = ()--()-->()--(node_3:label4), path_1 = ()-[edge_1*]-()<-[reltype7*0]-(node_3:labelunk) WHERE TRUE >= edge_1.a UNWIND [] AS z RETURN edge_1.d

------ CONFIG DEBUG OUTPUT ------
lazyfree-lazy-expire no
list-compress-depth 0
slave-read-only yes
lazyfree-lazy-user-flush no
io-threads-do-reads no
lazyfree-lazy-user-del no
repl-diskless-load disabled
sanitize-dump-payload no
repl-diskless-sync yes
io-threads 1
proto-max-bulk-len 512mb
lazyfree-lazy-server-del no
replica-read-only yes
client-query-buffer-limit 1gb
activedefrag no
lazyfree-lazy-eviction no

------ FAST MEMORY TEST ------
2590711:M 21 Jan 2023 03:45:56.627 # main thread terminated
2590711:M 21 Jan 2023 03:45:56.628 # Bio thread for job type #0 terminated
2590711:M 21 Jan 2023 03:45:56.628 # Bio thread for job type #1 terminated
2590711:M 21 Jan 2023 03:45:56.628 # Bio thread for job type #2 terminated

Fast memory test PASSED, however your memory can still be broken. Please run a memory test for several hours if possible.

------ DUMPING CODE AROUND EIP ------
Symbol: (null) (base: (nil))
Module: /lib/x86_64-linux-gnu/libc.so.6 (base 0x7ff87b682000)
$ xxd -r -p /tmp/dump.hex /tmp/dump.bin
$ objdump --adjust-vma=(nil) -D -b binary -m i386:x86-64 /tmp/dump.bin
------

=== REDIS BUG REPORT END. Make sure to include from START to END. ===

       Please report the crash by opening an issue on github:

           http://github.com/redis/redis/issues

  If a Redis module was involved, please open in the module's repo instead.

  Suspect RAM error? Use redis-server --test-memory to verify it.

  Some other issues could be detected by redis-server --check-system
[1]    2590711 segmentation fault (core dumped)  ./src/redis-server ./redis.conf
```

### PR DESCRIPTION
**Pull request by swilly22**
Resolves  #2836

### CODE REVIEW

File: src/arithmetic/algebraic_expression/algebraic_expression_construction.c [hunk 1]
```diff
@@ -329,81 +331,73 @@ static AlgebraicExpression *_AlgebraicExpression_OperandFromEdge
...
-		default: // multiple edge type: -[:A|:B]->
-			add = AlgebraicExpression_NewOperation(AL_EXP_ADD);
-			for(uint i = 0; i < reltype_count; i++) {
-				AlgebraicExpression *operand = AlgebraicExpression_NewOperand(
-						NULL, false, src, dest, edge, e->reltypes[i]);
-				AlgebraicExpression_AddChild(add, operand);
-			}
-			root = add;
-			break;
+	uint reltype_count = array_len(e->reltypeIDs);
```
Comment (AviAvni) [inline src/arithmetic/algebraic_expression/algebraic_expression_construction.c:351]: so now we don't use IDENTITY_MATRIX anymore maybe we can delete it?

File: src/execution_plan/execution_plan_build/build_match_op_tree.c [hunk 2]
```diff
@@ -109,10 +109,10 @@ static void _ExecutionPlan_ProcessQueryGraph
 			}
 
 			if(edge && (QGEdge_VariableLength(edge) || !QGEdge_SingleHop(edge))) {
-				// edge is part of a shortest-path
-				// MATCH allShortestPaths((a)-[*..]->(b))
-				// validate both edge ends are bounded
 				if(QGEdge_IsShortestPath(edge)) {
+					// edge is part of a shortest-path
```
Comment (AviAvni) [inline src/execution_plan/execution_plan_build/build_match_op_tree.c:113]: don't we miss check QGEdge_GhostEdge in line 111

**Review by AviAvni — COMMENTED** [review]
great ee comments

File: src/arithmetic/algebraic_expression/algebraic_expression_construction.c (hunk 1 above)
Comment (swilly22) [reply src/arithmetic/algebraic_expression/algebraic_expression_construction.c:351]: we sure can

File: src/execution_plan/execution_plan_build/build_match_op_tree.c (hunk 2 above)
Comment (swilly22) [reply src/execution_plan/execution_plan_build/build_match_op_tree.c:113]: no,
if `edge` is not NULL AND
`QGEdge_VariableLength(edge)` is `FALSE` AND
`!QGEdge_SingleHop(edge)` returns `TRUE`

This means we're dealing with a 0 length edge.

edge is referenced, not variable length and not single hop.
this means constant length which is not of length 1, could be 0, 2, 3....
in which case we need to use the `CondVarLenTraverseOp` operation

File: src/arithmetic/algebraic_expression/algebraic_expression_construction.c [hunk 3]
```diff
@@ -44,7 +44,9 @@ static inline bool _should_populate_edge
 	QGEdge *e
 ) {
 	ASSERT(e != NULL);
-	return (_referred_entity(e->alias) || QGEdge_VariableLength(e));
+	return (_referred_entity(e->alias) ||
+			QGEdge_VariableLength(e)   ||
+			QGEdge_GhostEdge(e));
```
Comment (swilly22) [inline src/arithmetic/algebraic_expression/algebraic_expression_construction.c:49]: can this be removed?

**Pull request merged**